romanisim/l1.py

```diff
@@ -282,48 +288,72 @@ def apportion_counts_to_resultants(
+
     if persistence is not None:
         tstart = tstart.mjd
 
+    # Loop over read probabilities
     for i, pi in enumerate(pij):
+        # Reset resultant counts
         resultant_counts[...] = 0
+
+        # Loop over resultant probabilities
         for j, p in enumerate(pi):
-            if linearity is None or ((i == 0) and (j == 0)):
-                ki = 1
-                efficiency = 1
-            else:
-                ki = (1 - ki_numerator[readnum - 1]) / (1 - ki_denominator)
-                efficiency = linearity.efficiency(
-                    counts_so_far
-                    + counts * (efficiency * pij_per_read[i][j] / 2))
-                m = (efficiency < 0) | (efficiency > 1)
-                nlflag[m] = True
-                efficiency = np.clip(efficiency, 0, 1)
-            read = rng_numpy.binomial((counts - counts_so_far).astype('i4'),
-                                      p * efficiency * ki)
+            # if linearity is None or ((i == 0) and (j == 0)):
+            #     ki = 1
+            #     efficiency = 1
+            # else:
+            #     ki = (1 - ki_numerator[readnum - 1]) / (1 - ki_denominator)
+            #     efficiency = linearity.efficiency(
+            #         counts_so_far
+            #         + counts * (efficiency * pij_per_read[i][j] / 2))
+            #     m = (efficiency < 0) | (efficiency > 1)
+            #     nlflag[m] = True
+            #     efficiency = np.clip(efficiency, 0, 1)
+            # read = rng_numpy.binomial((counts - counts_so_far).astype('i4'),
+            #                           p * efficiency * ki)
+
+            # Set read counts
+            read = rng_numpy.binomial((counts - counts_so_far).astype('i4'), p)
             counts_so_far += read
+
+            # Apply cosmic rays
             if crparam is not None:
                 old_instrumental_so_far = instrumental_so_far.copy()
                 cr.simulate_crs(instrumental_so_far, parameters.read_time,
                                 **crparam, rng=rng_numpy_cr)
                 crhits = instrumental_so_far != old_instrumental_so_far
                 dq[i, crhits] |= parameters.dqbits['jump_det']
+
+            # Apply persistence
             if persistence is not None:
                 tnow = tstart + tij[i][j] / (24 * 60 * 60)
                 persistence.add_to_read(
                     instrumental_so_far, tnow, rng=rng_numpy_ps)
+
+            # Update counts for the resultant
             resultant_counts += counts_so_far + instrumental_so_far
+            # if linearity is not None:
+            #     ki_denominator += efficiency * pij_per_read[i][j]
+            # readnum += 1
+
+            # Apply linearity
             if linearity is not None:
-                ki_denominator += efficiency * pij_per_read[i][j]
-            readnum += 1
+                resultant_counts += linearity.apply(counts_so_far + instrumental_so_far, electrons=True)
```

Comment:
This is applying the counts twice if linearity is not None.  This needs to be more like:
if not None:
    resultant_counts += linearity.apply(...)
else:
    resultant_counts += ...

Reply:
Oops.. line moved.